x-pack/auditbeat/module/system/socket/guess/guess.go

```diff
@@ -0,0 +1,296 @@
+}
+
+func guessOnce(guesser Guesser, installer helper.ProbeInstaller, ctx Context) (result common.MapStr, err error) {
+	if err := guesser.Prepare(ctx); err != nil {
+		return nil, errors.Wrap(err, "prepare failed")
+	}
+	defer func() {
+		if err := guesser.Terminate(); err != nil {
+			ctx.Log.Errorf("Terminate failed: %v", err)
+		}
+	}()
+	probes, err := guesser.Probes()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed generating probes")
+	}
+
+	decoders := make([]tracing.Decoder, 0, len(probes))
+	formats := make([]tracing.ProbeFormat, 0, len(probes))
+	defer installer.UninstallInstalled()
+	for _, pdesc := range probes {
+		format, decoder, err := installer.Install(pdesc)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to add kprobe '%s'", pdesc.Probe.String())
+		}
+		formats = append(formats, format)
+		decoders = append(decoders, decoder)
+	}
+
+	// channel to receive the TID and sync with the trigger goroutine.
+	// Length is zero so that it blocks sender and receiver, making it useful
+	// for synchronization.
+	tidChan := make(chan int, 0)
+
+	// Trigger goroutine.
+	go func() {
+		// Lock this goroutine to the current CPU thread. This way we can setup
+		// the perf channel to receive events from this thread only, avoiding
+		// the guess being contaminated with an externally-generated event.
+		runtime.LockOSThread()
+		defer runtime.UnlockOSThread()
+		defer close(tidChan)
+
+		// Send the thread ID
+		tidChan <- syscall.Gettid()
+
+		// This blocks until the main goroutine installs the perf channel
+		tidChan <- 0
+
+		if err := guesser.Trigger(); err != nil {
+			ctx.Log.Errorf("Trigger failed (%s): %v", guesser.Name(), err)
+		}
+	}()
+
+	// Receive the thread ID from the trigger goroutine.
+	tid := <-tidChan
+	defer func() {
+		// reads the tidChan just in case we return due to an error below and
+		// the trigger goroutine is blocking for the fire signal.
+		// Otherwise this will be a read from a closed channel.
+		select {
+		case <-tidChan:
+		default:
+		}
+	}()
+	perfchan, err := tracing.NewPerfChannel(
+		tracing.WithBufferSize(8),
+		tracing.WithErrBufferSize(1),
+		tracing.WithLostBufferSize(8),
+		tracing.WithRingSizeExponent(2),
+		tracing.WithTID(tid),
+		tracing.WithPollTimeout(time.Millisecond*10))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create perfchannel")
+	}
+	defer perfchan.Close()
+
+	for i := range probes {
+		if err := perfchan.MonitorProbe(formats[i], decoders[i]); err != nil {
+			return nil, errors.Wrap(err, "failed to monitor probe")
+		}
+	}
+
+	if err := perfchan.Run(); err != nil {
+		return nil, errors.Wrap(err, "failed to run perf channel")
+	}
+
+	timer := time.NewTimer(ctx.Timeout)
+
+	defer func() {
+		if !timer.Stop() {
+			select {
+			case <-timer.C:
+			default:
+			}
+		}
+	}()
+
+	// Read the tidchan again to release the trigger.
+	<-tidChan
+
+	// This blocks until trigger terminates (channel closed) or a timeout
+	select {
+	case <-tidChan:
```

Comment:
Actually, do we need to wait here for trigger to finish? Wouldn't `perfchan.C()` wait till `Trigger` provokes the events?

Reply:
It needs to make sure that Validate() and Trigger() don't run at the same time, as the guesses are not thread-safe.

For example in some cases Trigger() might generate the kprobe event (call a syscall) and then update some shared state that Validate() needs (i.e. number of bytes written).

Reply:
Oh, right, then probably the other comment won't work.